server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/PartitionCache.java

```diff
@@ -0,0 +1,757 @@
+   * @param storageGroupToDeviceMap the result map that match storage group to device, return empty
+   *     when miss
+   * @param failedDevices the devices that failed to match
+   * @param failFast if true, return when failed. if false, return when all devices finish match
+   * @return whether hit
+   */
+  private boolean getStorageGroupToDeviceMap(
+      List<String> devicePaths,
+      Map<String, List<String>> storageGroupToDeviceMap,
+      List<String> failedDevices,
+      boolean failFast) {
+    storageGroupCacheLock.readLock().lock();
+    try {
+      boolean result = true;
+      for (String devicePath : devicePaths) {
+        String storageGroup = getStorageGroup(devicePath);
+        if (null == storageGroup) {
+          logger.debug(
+              "[{} Cache] miss when search device {}", STORAGE_GROUP_CACHE_NAME, devicePath);
+          result = false;
+          if (failFast) {
+            break;
+          } else {
+            failedDevices.add(devicePath);
+          }
+        } else {
+          List<String> devices =
+              storageGroupToDeviceMap.computeIfAbsent(storageGroup, k -> new ArrayList<>());
+          devices.add(devicePath);
+        }
+      }
+      if (!result) {
+        storageGroupToDeviceMap.clear();
+      }
+      logger.debug("[{} Cache] hit when search device {}", STORAGE_GROUP_CACHE_NAME, devicePaths);
+      CacheMetricsRecorder.record(result, STORAGE_GROUP_CACHE_NAME);
+      return result;
+    } finally {
+      storageGroupCacheLock.readLock().unlock();
+    }
+  }
+
+  /**
+   * update storage group cache
+   *
+   * @param storageGroupNames the storage groups that need to update
+   */
+  public void updateStorageCache(Set<String> storageGroupNames) {
+    storageGroupCacheLock.writeLock().lock();
+    try {
+      storageGroupCache.addAll(storageGroupNames);
+    } finally {
+      storageGroupCacheLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * invalid storage group cache
+   *
+   * @param storageGroupNames the storage groups that need to invalid
+   */
+  public void invalidStorageGroupCache(List<String> storageGroupNames) {
+    storageGroupCacheLock.writeLock().lock();
+    try {
+      for (String storageGroupName : storageGroupNames) {
+        storageGroupCache.remove(storageGroupName);
+      }
+    } finally {
+      storageGroupCacheLock.writeLock().unlock();
+    }
+  }
+
+  /** invalid all storage group cache */
+  public void invalidAllStorageGroupCache() {
+    storageGroupCacheLock.writeLock().lock();
+    try {
+      storageGroupCache.clear();
+    } finally {
+      storageGroupCacheLock.writeLock().unlock();
+    }
+  }
+
+  // endregion
+
+  // region replicaSet cache
+  /** get regionReplicaSet from consensusGroupId */
+  public TRegionReplicaSet getRegionReplicaSet(TConsensusGroupId consensusGroupId) {
+    TRegionReplicaSet regionReplicaSet;
+    if (!groupIdToReplicaSetMap.containsKey(consensusGroupId)) {
```

Comment:
Need to handle concurrent issue here. The sample code should be

```
if (! xxx) {
   lock(cache) {
     if (! xxx) {
        fetchGroupIdToReplicaSet
     }
   }
}
```

Reply:
Get it